File: src/javascript/sorting/6.heapSort/heapSort.js
function heapSort(arr) {
    let size = arr.length;

    for (let i = Math.floor(size/2 - 1); i >= 0; i--) {
        heapify(arr, size, i);        
    }

    for (let i = size - 1; i >= 0; i--) {
        [arr[i], arr[0]] = [arr[0], arr[i]];
        heapify(arr, i, 0);
    }
}

function heapify(arr, size, parentIndex) {
    let largest = parentIndex;
    let leftChild = 2 * parentIndex + 1;
    let rightChild = 2 * parentIndex + 2;

    if(leftChild < size && arr[leftChild] > arr[largest]) {
        largest = leftChild;
    }

    if(rightChild < size && arr[rightChild] > arr[largest]) {
        largest = rightChild;
    }

    if(parentIndex !== largest) {
        [arr[parentIndex], arr[largest]] = [arr[largest], arr[parentIndex]];
        heapify(arr, size, largest);
    }
}

let arr = [40, 50, 20, 0, -10, 30, 10];
heapSort(arr);
console.log(arr);

File: src/python/sorting/6.heapSort/heap_sort.py
"""
Heap Sort Algorithm

Time Complexity: O(n log n)
Space Complexity: O(1)

Heap sort uses a binary heap data structure to sort elements.
It first builds a max heap, then repeatedly extracts the maximum element.
"""


def heap_sort(arr):
    """
    Sorts an array using heap sort algorithm.
    
    Args:
        arr: List of comparable elements
    """
    size = len(arr)

    # Build max heap
    for i in range(size // 2 - 1, -1, -1):
        heapify(arr, size, i)

    # Extract elements from heap one by one
    for i in range(size - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # Swap
        heapify(arr, i, 0)


def heapify(arr, size, parent_index):
    """
    Maintains the heap property for a subtree.
    
    Args:
        arr: Array representing the heap
        size: Size of the heap
        parent_index: Index of the parent node
    """
    largest = parent_index
    left_child = 2 * parent_index + 1
    right_child = 2 * parent_index + 2

    if left_child < size and arr[left_child] > arr[largest]:
        largest = left_child

    if right_child < size and arr[right_child] > arr[largest]:
        largest = right_child

    if parent_index != largest:
        arr[parent_index], arr[largest] = arr[largest], arr[parent_index]
        heapify(arr, size, largest)


# Usage example
if __name__ == "__main__":
    arr = [40, 50, 20, 0, -10, 30, 10]
    print(f"Original: {arr}")
    heap_sort(arr)
    print(f"Sorted: {arr}")
